### PR TITLE
Toggleable orthographic camera added.

### PR DESCRIPTION
**Pull request by aemreaydin**
Fix: #17 #1096

### CODE REVIEW

File: library/VTKExtensions/Rendering/vtkF3DRenderer.cxx [hunk 1]
```diff
@@ -1242,6 +1246,7 @@ void vtkF3DRenderer::ConfigureCheatSheet()
     cheatSheetText << " 3: Right View camera\n";
     cheatSheetText << " 7: Top View camera\n";
     cheatSheetText << " 9: Isometric View camera\n";
+    cheatSheetText << " 0: Toggle Orthogonal projection\n";
```
Comment (mwestphal) [inline library/VTKExtensions/Rendering/vtkF3DRenderer.cxx:1249]: @lknknm suggested 5 in the issue: https://github.com/f3d-app/f3d/issues/885

Any reason to change that ?

File: library/VTKExtensions/Rendering/vtkF3DRenderer.cxx [hunk 2]
```diff
@@ -1276,6 +1281,17 @@ void vtkF3DRenderer::ShowHDRISkybox(bool show)
...
 }
 
+void vtkF3DRenderer::ShowOrthogonal(bool show)
+{
+  if(this->OrthogonalVisible != show)
+  {
+    std::string orthogonalText = "Projection: Orthogonal";
+    this->OrthogonalVisible = show;
+    this->OrthogonalActor->SetText(vtkCornerAnnotation::UpperLeft, orthogonalText.c_str());
+    this->OrthogonalActor->SetVisibility(show);
```
Comment (mwestphal) [inline library/VTKExtensions/Rendering/vtkF3DRenderer.cxx:1291]: no need for an text actor, the cheatsheet will inform the user

File: library/VTKExtensions/Rendering/vtkF3DRenderer.cxx (hunk 1 above)
Comment (aemreaydin) [reply library/VTKExtensions/Rendering/vtkF3DRenderer.cxx:1249]: simple wasnt aware of this ticket. The ticket I looked at was #1096.

Comment (mwestphal) [reply library/VTKExtensions/Rendering/vtkF3DRenderer.cxx:1249]: no worries, I should have pointed it out

**Comment by snoyer** [comment]
Any consensus on "orthographic" vs "orthogonal" vs "parallel" to name the projection within the app/UI?

Wikipedia says _"Orthographic projection (also orthogonal projection and analemma) [...] is a form of parallel projection"_ and has [a nice illustration](https://commons.wikimedia.org/wiki/File:Comparison_of_graphical_projections.svg) which is very educational but doesn't solve the choice.

**Comment by Meakk** [comment]
I vote for Orthographic.

File: .gitignore [hunk 3]
```diff
@@ -1,4 +1,5 @@
 .vscode
+.idea
```
Comment (mwestphal) [inline .gitignore:2]: curious, what is a .idea file ?

File: doc/user/OPTIONS.md [hunk 4]
```diff
@@ -88,16 +88,17 @@ Options|Default|Description
...
+------|---------|------
+\-\-camera-position=\<X,Y,Z\>|         |Set the camera position, overrides --camera-direction and camera-zoom-factor.
+\-\-camera-focal-point=\<X,Y,Z\>|         |Set the camera focal point.
+\-\-camera-view-up=\<X,Y,Z\>|         |Set the camera view up vector. Will be orthogonalized.
+\-\-camera-view-angle=\<angle\>|         |Set the camera view angle, a strictly positive value in degrees.
+\-\-camera-direction=\<X,Y,Z\>|         |Set the camera direction, looking at the focal point.
+\-\-camera-zoom-factor=\<factor\>|         |Set the camera zoom factor relative to the autozoom on data, a strictly positive value.
+\-\-camera-azimuth-angle=\<angle\>| 0.0     |Apply an azimuth transformation to the camera, in degrees, added after other camera options.
+\-\-camera-elevation-angle=\<angle\>| 0.0     |Apply an elevation transformation to the camera, in degrees, added after other camera options.
+\-\-camera-orthographic|         |Enable orthographic projection.
```
Comment (mwestphal) [inline doc/user/OPTIONS.md:101]: any specific reason to change the spacing ?

File: library/VTKExtensions/Rendering/vtkF3DRenderer.cxx [hunk 5]
```diff
@@ -1334,6 +1335,22 @@ void vtkF3DRenderer::ShowEdge(bool show)
...
+void vtkF3DRenderer::SetUseOrthographicProjection(bool use)
+{
+  vtkCamera* camera = GetActiveCamera();
+  if (use)
+  {
+    const double* position = camera->GetPosition();
+    const double* focal = camera->GetFocalPoint();
+    const double angle = camera->GetViewAngle();
+    double distance = std::sqrt(vtkMath::Distance2BetweenPoints(position, focal));
+    double parallel_scale = distance * tan(0.5 * vtkMath::RadiansFromDegrees(angle));
```
Comment (mwestphal) [inline library/VTKExtensions/Rendering/vtkF3DRenderer.cxx:1348]: use camelCase

File: library/src/camera_impl.cxx [hunk 6]
```diff
@@ -264,4 +264,5 @@ vtkCamera* camera_impl::GetVTKCamera()
 {
   return this->Internals->VTKRenderer->GetActiveCamera();
 }
-};
+
+}
```
Comment (mwestphal) [inline library/src/camera_impl.cxx:268]: missing eol

File: library/src/interactor_impl.cxx [hunk 7]
```diff
@@ -101,7 +102,7 @@ class interactor_impl::internals
     VT_FRONT,
     VT_RIGHT,
     VT_TOP,
-    VT_ISOMETRIC
+    VT_ISOMETRIC,
```
Comment (mwestphal) [inline library/src/interactor_impl.cxx:105]: should not be needed ?

File: library/src/interactor_impl.cxx [hunk 8]
```diff
@@ -319,6 +320,11 @@ class interactor_impl::internals
         self->SetViewOrbit(ViewType::VT_RIGHT, self);
         render = true;
         break;
+      case '5':
+        self->Options.toggle("camera.orthographic");
+        ren->SetUseOrthographicProjection(self->Options.getAsBool("camera.orthographic"));
```
Comment (mwestphal) [inline library/src/interactor_impl.cxx:325]: chaning the option should be enough

File: library/src/interactor_impl.cxx [hunk 9]
```diff
@@ -486,10 +492,7 @@ class interactor_impl::internals
   std::function<bool(const std::vector<std::string>&)> DropFilesUserCallBack =
     [](const std::vector<std::string>&) { return false; };
 
-  void StartInteractor()
-  {
-    this->VTKInteractor->Start();
-  }
+  void StartInteractor() { this->VTKInteractor->Start(); }
```
Comment (mwestphal) [inline library/src/interactor_impl.cxx:495]: not needed until we update the cland format

**Review by mwestphal — CHANGES_REQUESTED** [review]

File: .gitignore (hunk 3 above)
Comment (aemreaydin) [reply .gitignore:2]: It's just coming from Jetbrains IDE's basically the same as .vscode, ide settings.

File: doc/user/OPTIONS.md (hunk 4 above)
Comment (aemreaydin) [reply doc/user/OPTIONS.md:101]: IDE auto formatted it seems like

File: library/src/camera_impl.cxx (hunk 6 above)
Comment (aemreaydin) [reply library/src/camera_impl.cxx:268]: will update

File: library/src/interactor_impl.cxx (hunk 7 above)
Comment (aemreaydin) [reply library/src/interactor_impl.cxx:105]: will update

File: library/src/interactor_impl.cxx (hunk 8 above)
Comment (aemreaydin) [reply library/src/interactor_impl.cxx:325]: you're right, updating.

File: library/VTKExtensions/Rendering/vtkF3DRenderer.h [hunk 10]
```diff
@@ -77,6 +77,13 @@ class vtkF3DRenderer : public vtkOpenGLRenderer
   void SetRaytracingSamples(int samples);
   ///@}
 
+  ///@{
```
Comment (Meakk) [inline library/VTKExtensions/Rendering/vtkF3DRenderer.h:80]: doxygen grouping tags are not required for a single function

**Review by Meakk — CHANGES_REQUESTED** [review]
Here's the list of failing tests. They are all related to the cheatsheet, and it makes sense. Please update the baselines.

```
The following tests FAILED:
	165 - f3d::TestInteractionCheatsheetRaytracing (Failed)
	166 - f3d::TestInteractionCheatsheetWhiteBGRaytracing (Failed)
	167 - f3d::TestInteractionCheatsheetBlackBGRaytracing (Failed)
	168 - f3d::TestInteractionCheatsheetScalarsRaytracing (Failed)
	169 - f3d::TestInteractionNoFileCheatsheetRaytracing (Failed)
Errors while running CTest
```

**Comment by Meakk** [comment]
[baselines.zip](https://github.com/f3d-app/f3d/files/13708032/baselines.zip)

Just in case you need the new baselines

**Comment by mwestphal** [comment]
@aemreaydin please "resolve" discussion that you have adressed :)

**Comment by mwestphal** [comment]
You need to rebase again. I would suggest to add changelog right before mergingn to avoid this :)

**Comment by mwestphal** [comment]
Hi @aemreaydin , need any help moving forward ?

**Comment by mwestphal** [comment]
Hi @aemreaydin , please let us know if you still want to work on this :)

**Comment by mwestphal** [comment]
superseeded by https://github.com/f3d-app/f3d/pull/1235

**Pull request closed without merging**